src/sentry/unmerge.py

```diff
@@ -0,0 +1,215 @@
+        fingerprints: Sequence[str],
+        actor_id: Optional[int],
+        last_event: Optional[str] = None,
+        batch_size: int = 500,
+        source_fields_reset: bool = False,
+        eventstream_state: Any = None,
+        replacement: Optional[UnmergeReplacement] = None,
+        locked_primary_hashes: Optional[Collection[str]] = None,
+        destinations: Optional[Destinations] = None,
+    ) -> "UnmergeArgs":
+        if destinations is None:
+            if destination_id is not None:
+                destinations = {_DEFAULT_UNMERGE_KEY: (destination_id, eventstream_state)}
+            else:
+                destinations = {}
+
+        if last_event is None:
+            assert eventstream_state is None
+            assert not source_fields_reset
+
+            return InitialUnmergeArgs(
+                project_id=project_id,
+                source_id=source_id,
+                replacement=UnmergeReplacement.parse_arguments(fingerprints, replacement),
+                actor_id=actor_id,
+                batch_size=batch_size,
+                destinations=destinations,
+            )
+        else:
+            assert locked_primary_hashes is not None or fingerprints is not None
+            return SuccessiveUnmergeArgs(
+                project_id=project_id,
+                source_id=source_id,
+                replacement=UnmergeReplacement.parse_arguments(fingerprints, replacement),
+                actor_id=actor_id,
+                batch_size=batch_size,
+                last_event=last_event,
+                destinations=destinations,
+                locked_primary_hashes=locked_primary_hashes or fingerprints or [],
+                source_fields_reset=source_fields_reset,
+            )
+
+    def dump_arguments(self) -> Mapping[str, Any]:
+        rv = dataclasses.asdict(self)
+        rv["fingerprints"] = None
+        rv["destination_id"] = None
+        rv["replacement"] = self.replacement
+        return rv
+
+
+@dataclass(frozen=True)
+class InitialUnmergeArgs(UnmergeArgsBase):
+    # In tests the destination task is passed in explicitly from the outside,
+    # so we support unmerging into an existing destination group. In production
+    # this does not happen.
+    destinations: InitialDestinations
+
+
+@dataclass(frozen=True)
+class SuccessiveUnmergeArgs(UnmergeArgsBase):
+    last_event: Optional[Any]
+    locked_primary_hashes: Collection[str]
+
+    # unmerge may only start mutating data on a successive page, once it
+    # actually has found an event that needs to be migrated.
+    # (unmerge_key) -> (group_id, eventstream_state)
+    destinations: Destinations
+
+    # likewise unmerge may only find "source" events (events that should not be
+    # migrated) on the second page, only then (and only once) it can reset
+    # group attributes such as last_seen.
+    source_fields_reset: bool
+
+
+UnmergeArgs = Union[InitialUnmergeArgs, SuccessiveUnmergeArgs]
```

Comment:
Why do you need to define this type? Isn't it the same as `UnmergeArgsBase`?

Reply:
It exhaustively enumerates all possible types for type inference

Then you can say:

```
if isinstance(args, InitialUnmergeArgs):
    ...
else:
    ...
```

and the else-branch will have the attributes of SuccessiveUnmergeArgs available

Reply:
Yeah that's true. It would need to be:
```
if isinstance(args, InitialUnmergeArgs):
    ...
else:
    assert isinstance(args, SuccessiveUnmergeArgs)
    ...
```